sway/layout.c

```diff
@@ -895,136 +955,322 @@ static void arrange_windows_r(swayc_t *container, double width, double height) {
-			// depend on the title bar geometry of its siblings.
-			if (focused && container->children->length > 1) {
-				update_container_border(focused);
+			if (i == end - 1) {
+				double remaining_height = y + height - child_y;
+				arrange_windows_r(child, width, remaining_height);
+			} else {
+				arrange_windows_r(child, width, child->height * scale);
 			}
+			child_y += child->height;
+		}
+
+		// update focused view border last because it may
+		// depend on the title bar geometry of its siblings.
+		if (focused && container->children->length > 1) {
+			update_container_border(focused);
+		}
+	}
+}
+
+void apply_tabbed_or_stacked_layout(swayc_t *container, double x, double y,
+				    double width, double height) {
+	int i;
+	swayc_t *focused = NULL;
+	for (i = 0; i < container->children->length; ++i) {
+		swayc_t *child = container->children->items[i];
+		child->x = x;
+		child->y = y;
+		if (child == container->focused) {
+			focused = child;
+		} else {
+			arrange_windows_r(child, width, height);
 		}
+	}
+
+	if (focused) {
+		arrange_windows_r(focused, width, height);
+	}
+}
+
+void apply_auto_layout(swayc_t *container, const double x, const double y,
+		       const double width, const double height,
+		       enum swayc_layouts group_layout,
+		       bool master_first) {
+	// Auto layout "container" in width x height @ x, y
+	// using "group_layout" for each of the groups in the container.
+	// There is one "master" group, plus container->nb_slave_groups.
+	// Each group is layed out side by side following the "major" axis.
+	// The direction of the layout used for groups is the "minor" axis.
+	// Example:
+	//
+	//     ---- major axis -->
+	//   +---------+-----------+
+	//   |         |           |   |
+	//   | master  | slave 1   |   |
+	//   |         +-----------+   | minor axis (direction of group_layout)
+	//   |         |           |   |
+	//   |         | slave 2   |   V
+	//   +---------+-----------+
+	//
+	//  container with three children (one master and two slaves) and
+	//  a single slave group (containing slave 1 and 2). The master
+	//  group and slave group are layed out using L_VERT.
+
+	uint_fast32_t nb_slaves = container->children->length - container->nb_master;
+	uint_fast32_t nb_groups = (container->nb_master > 0 ? 1 : 0) +
+		MIN(container->nb_slave_groups, nb_slaves);
+
+	// the target dimension of the container along the "major" axis, each
+	// group in the container will be layed out using "group_layout" along
+	// the "minor" axis.
+	double dim_maj;
+	double pos_maj;
+
+	// x and y coords for the next group to be laid out.
+	const double *group_x, *group_y;
+
+	// pos of the next group to layout along the major axis
+	double pos;
+
+	// size of the next group along the major axis.
+	double group_dim;
+
+	// height and width of next group to be laid out.
+	const double *group_h, *group_w;
+
+	switch(group_layout) {
+	default:
+		sway_log(L_ERROR, "Unknown layout type (%d) used in %s()",
```

Comment:
Make this L_DEBUG. It's programmer error, right? Not user error? And is a recoverable condition?